docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/launcher/DockerComputerJNLPLauncher.java

```diff
@@ -0,0 +1,214 @@
+                TimeUnit2.SECONDS.toMillis(launchTimeout) > System.currentTimeMillis() - launchTime) {
+            logger.println("Waiting slave connection...");
+            Thread.sleep(1000);
+        }
+
+        if (!dockerComputer.isOnline()) {
+            LOGGER.info("Launch timeout, termintaing slave based on '{}'", containerId);
+            logger.println("Launch timeout, termintaing slave.");
+            dockerComputer.getNode().terminate();
+            throw new IOException("Can't connect slave to jenkins");
+        }
+
+        LOGGER.info("Launched slave '{}' based on '{}'", dockerComputer.getName(), containerId);
+        logger.println("Launched slave for " + containerId);
+    }
+
+    @Override
+    public ComputerLauncher getPreparedLauncher(String cloudId, DockerTemplate template, InspectContainerResponse containerInspectResponse) {
+        return new DockerComputerJNLPLauncher(getJnlpLauncher());
+    }
+
+    @Override
+    public void appendContainerConfig(DockerTemplate dockerTemplate, CreateContainerCmd createContainerCmd) throws IOException {
+
+        try (InputStream istream = DockerComputerJNLPLauncher.class.getResourceAsStream("DockerComputerJNLPLauncher/init.sh")) {
+            final String initCmd = IOUtils.toString(istream, Charsets.UTF_8);
+            if (initCmd == null) {
+                throw new IllegalStateException("Resource file 'init.sh' not found");
+            }
+//            createContainerCmd.withCmd("/bin/sh"); // nop
+            // wait for params
+            createContainerCmd.withCmd("/bin/bash",
+                    "-cxe",
+                    "cat << EOF >> /tmp/init.sh && chmod +x /tmp/init.sh && exec /tmp/init.sh\n" +
+                            initCmd.replace("$", "\\$") + "\n" +
+                            "EOF" + "\n"
+            );
+        }
+
+//        final String homeDir = dockerTemplate.getRemoteFs();
+//        if (isNotBlank(homeDir)) {
+//            createContainerCmd.withWorkingDir(homeDir);
+//        }
+//
+//        final String user = dockerTemplate.getUser();
+//        if (isNotBlank(user)) {
+//            createContainerCmd.withUser(user);
+//        }
+
+        createContainerCmd.withTty(true);
+        createContainerCmd.withStdinOpen(true);
+
+    }
+
+    @Override
+    public boolean waitUp(String cloudId, DockerTemplate dockerTemplate, InspectContainerResponse ir) {
+        return super.waitUp(cloudId, dockerTemplate, ir);
+    }
+
+    @Override
+    public Descriptor<ComputerLauncher> getDescriptor() {
+        return DESCRIPTOR;
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
+
+    public static class DescriptorImpl extends Descriptor<ComputerLauncher> {
+
+        public Class getJNLPLauncher() {
+            return JNLPLauncher.class;
+        }
+
+        @Override
+        public String getDisplayName() {
+            return "(Experimental) Docker JNLP launcher";
```

Comment:
@steve-jansen i'm marking this launcher as experimental. And if it doesn't work - will fix later. I think everybody need fix related to ssh. 
Or even will remove from available for some time...